File: backend/app/schemas/backtest.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class BacktestMetrics(BaseModel):
    total_return: float = Field(description="Cumulative portfolio return")
    benchmark_total_return: float = Field(description="Cumulative equal-weight universe return")
    max_drawdown: float = Field(description="Maximum drawdown (positive fraction)")
    sharpe_ratio: float | None = Field(description="Annualized Sharpe ratio")
    periods: int = Field(description="Number of rebalance periods")
    win_rate: float = Field(description="Fraction of periods with positive return")


class BacktestEquityPoint(BaseModel):
    date: date
    portfolio: float
    benchmark: float


class BacktestPeriodRow(BaseModel):
    rebalance_date: date
    next_date: date
    holdings: list[str]
    period_return: float
    benchmark_return: float


class BacktestResponse(BaseModel):
    market: str
    top_n: int
    start_date: date | None
    end_date: date | None
    snapshot_dates_used: list[date]
    metrics: BacktestMetrics
    equity_curve: list[BacktestEquityPoint]
    periods: list[BacktestPeriodRow]
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/backtest.ts
export interface BacktestMetrics {
  total_return: number;
  benchmark_total_return: number;
  max_drawdown: number;
  sharpe_ratio: number | null;
  periods: number;
  win_rate: number;
}

export interface BacktestEquityPoint {
  date: string;
  portfolio: number;
  benchmark: number;
}

export interface BacktestPeriodRow {
  rebalance_date: string;
  next_date: string;
  holdings: string[];
  period_return: number;
  benchmark_return: number;
}

export interface BacktestResponse {
  market: string;
  top_n: number;
  start_date: string | null;
  end_date: string | null;
  snapshot_dates_used: string[];
  metrics: BacktestMetrics;
  equity_curve: BacktestEquityPoint[];
  periods: BacktestPeriodRow[];
  warnings: string[];
}

export interface BacktestQueryParams {
  market?: string;
  top_n?: number;
  start_date?: string;
  end_date?: string;
}
